tests/functional/test_adcm_upgrade.py

```diff
@@ -160,3 +172,430 @@ def test_actions_availability_after_upgrade(
+                'providers': tuple(simple_providers),
+                'hosts': tuple(simple_hosts),
+                'clusters': tuple(simple_clusters),
+                'provider_bundle': simple_provider_bundle,
+                'cluster_bundle': simple_cluster_bundle,
+            },
+            'complex': {
+                'providers': {'host_supplier': complex_objects[0], 'free_hosts': complex_objects[1]},
+                'clusters': {
+                    'all_services': complex_objects[2],
+                    'config_history': complex_objects[3],
+                    'not_configured': complex_objects[4],
+                },
+            },
+            'upgrade': {'upgraded': upgraded_cluster, 'not_upgraded': not_upgraded_cluster},
+        }
+
+    # Test itself
+
+    @pytest.mark.full()
+    @pytest.mark.parametrize("adcm_is_upgradable", [True], indirect=True)
+    @pytest.mark.parametrize("image", [previous_adcm_version_tag()], indirect=True)
+    def test_upgrade_dirty_adcm(
+        self,
+        adcm_fs: ADCM,
+        sdk_client_fs: ADCMClient,
+        adcm_api_credentials: dict,
+        upgrade_target: Tuple[str, str],
+        dirty_adcm: dict,
+    ):
+        """
+        Create previous version ADCM with a lot of different objects.
+        Upgrade ADCM.
+        Run actions on ADCM.
+        """
+        objects_are_not_changed = build_objects_checker(changed=None, extractor=_get_object_fields)
+        with allure.step('Upgrade ADCM and expect all objects to be same'), objects_are_not_changed(
+            sdk_client_fs
+        ), self.check_job_related_objects_are_not_changed(sdk_client_fs):
+            _upgrade_adcm(adcm_fs, sdk_client_fs, adcm_api_credentials, upgrade_target)
+        self.run_actions_after_upgrade(
+            dirty_adcm['complex']['clusters']['all_services'],
+            dirty_adcm['complex']['clusters']['config_history'],
+            dirty_adcm['simple']['providers'][0],
+        )
+
+    # Steps and helpers
+
+    @contextmanager
+    def check_job_related_objects_are_not_changed(self, adcm_client: ADCMClient):
+        """Freeze jobs and check that they aren't changed after upgrade"""
+
+        def extract_job_info(job: Job) -> dict:
+            return {
+                'task_id': job.task_id,
+                'status': job.status,
+                'start_date': job.start_date,
+                'finish_date': job.finish_date,
+                'log_ids': {log.id for log in job.log_list()},
+            }
+
+        comparator = build_objects_comparator(
+            get_compare_value=extract_job_info, field_name='Job info', name_composer=lambda obj: f"Job with id {obj.id}"
+        )
+        jobs: List[Job] = get_objects_via_pagination(adcm_client.job_list)
+        frozen_objects = {job.job_id: extract_job_info(job) for job in jobs}
+
+        yield
+
+        for job_id, job_info in frozen_objects.items():
+            comparator(adcm_client.job(id=job_id), job_info)
```

Comment:
wrap steps like `Assert that Job with id 78 has ...` into step

Reply:
wrapped